src/java.base/share/classes/jdk/internal/misc/CarrierThread.java

```diff
@@ -60,27 +66,43 @@ public CarrierThread(ForkJoinPool pool) {
     }
 
     /**
-     * For use by {@link Blocker} to test if the thread is in a blocking operation.
+     * Mark the start of a blocking operation.
      */
-    boolean inBlocking() {
-        //assert JLA.currentCarrierThread() == this;
-        return blocking;
-    }
+    public void beginBlocking() {
+        assert Thread.currentThread().isVirtual() && JLA.currentCarrierThread() == this;
+        if (compensating == NOT_COMPENSATING) {
+            compensating = COMPENSATE_IN_PROGRESS;
 
-    /**
-     * For use by {@link Blocker} to mark the start of a blocking operation.
-     */
-    void beginBlocking() {
-        //assert JLA.currentCarrierThread() == this && !blocking;
-        blocking = true;
+            // FJP.tryCompensate to create or re-activate a spare thread
+            long value;
+            try {
+                value = ForkJoinPools.beginCompensatedBlock(getPool());
+            } catch (Throwable e) {
+                if (compensating == COMPENSATE_IN_PROGRESS) {
```

Comment:
> In the context of re-entrancy, if I am reading the code correctly, I don't see how a re-entrant call would end up on this line (and other similar lines) in this top level if block. When a thread enters the top level if block it immediately sets the compensating to COMPENSATE_IN_PROGRESS: I feel like I am missing something in this code.

These are fields on the carrier. If a virtual thread is preempted when compensating (or in the progress of) then it may continue on a different carrier or the original carrier may execute the task for a different virtual thread that does I/O. These are the cases that are handled here.